lib/auth/register.go

```diff
@@ -109,37 +109,48 @@ func Register(params RegisterParams) (*Identity, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	// Attempt to register through the auth server, if it fails, try and
-	// register through the proxy server.
-	ident, err := registerThroughAuth(token, params)
-	if err != nil {
-		// If no params client was set this is a proxy and fail right away.
-		if params.GetHostCredentials == nil {
-			log.Debugf("Missing client, failing with error from Auth Server: %v.", err)
-			return nil, trace.Wrap(err)
-		}
-
-		log.Errorf("Failed to register through auth server: %v; falling back to trying the proxy server", err)
+	type registerMethod struct {
+		call func(token string, params RegisterParams) (*Identity, error)
+		desc string
+	}
+	registerThroughAuth := registerMethod{registerThroughAuth, "to auth server"}
+	registerThroughProxy := registerMethod{registerThroughProxy, "through proxy server"}
+
+	var registerMethods []registerMethod
+	if params.GetHostCredentials == nil {
+		log.Debugf("Missing client, it is not possible to register through proxy.")
+		registerMethods = []registerMethod{registerThroughAuth}
+	} else if params.Servers[0].Port(0) == defaults.HTTPListenPort {
```

Comment:
Is there any validation of `RegisterParams` taking place prior to this? How do we know `params.Servers` is valid?
I'd even wrap this in a readable query:
```go
} else if authServerIsProxy(params) {
```

Reply:
You seem to be correct that `params.Servers` shouldn't be assumed to be valid at this point since `registerThroughProxy` actually checks `len(params.Servers)` as well. Updated the code, thanks!